### PR TITLE
[API] Initial checkout implementation

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | Based on #11666 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        StateMachineFactoryInterface $stateMachineFactory
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->customerFactory = $customerFactory;
+        $this->manager = $manager;
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    public function __invoke(AddressOrder $addressOrder): OrderInterface
+    {
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:53]: `$addressOrder->orderTokenValue` is used in three places, it should be assigned to new variable

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/ChoosePaymentMethodHandler.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        $stateMachine = $this->stateMachineFactory->get($cart, OrderCheckoutTransitions::GRAPH);
+
+        Assert::true($stateMachine->can(OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT), 'Order cannot have payment method assigned.');
+
+        /** @var PaymentMethodInterface $paymentMethod */
+        $paymentMethod = $this->paymentMethodRepository->findOneBy(['code' => $choosePaymentMethod->paymentMethod]);
+
+        Assert::notNull($paymentMethod, 'Payment method has not been found');
+        Assert::true(isset($cart->getPayments()[$choosePaymentMethod->paymentIdentifier]), 'Can not find payment with given identifier.');
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/ChoosePaymentMethodHandler.php:61]: `$choosePaymentMethod->paymentIdentifier` should be assigned to new variable

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/ChooseShippingMethodHandler.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            $stateMachine->can(OrderCheckoutTransitions::TRANSITION_SELECT_SHIPPING),
+            'Order cannot have shipment method assigned.'
+        );
+
+        /** @var ShippingMethodInterface $shippingMethod */
+        $shippingMethod = $this->shippingMethodRepository->findOneBy(['code' => $chooseShippingMethod->shippingMethod]);
+
+        Assert::notNull($shippingMethod, 'Shipping method has not been found');
+        Assert::true(
+            isset($cart->getShipments()[$chooseShippingMethod->shipmentIdentifier]),
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/ChooseShippingMethodHandler.php:71]: `$chooseShippingMethod->shipmentIdentifier` the same like above comment

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/CompleteOrderHandler.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        FactoryInterface $stateMachineFactory
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    public function __invoke(CompleteOrder $completeOrder): OrderInterface
+    {
+        /** @var OrderInterface $cart */
+        $cart = $this->orderRepository->findOneBy(['tokenValue' => $completeOrder->orderTokenValue]);
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/CompleteOrderHandler.php:42]: `$completeOrder->orderTokenValue` and here

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChoosePaymentMethodHandlerSpec.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        $this->beConstructedWith($orderRepository, $paymentMethodRepository, $stateMachineFactory);
+    }
+
+    function it_assignes_choosen_payment_method_to_specified_payment(
+        OrderRepositoryInterface $orderRepository,
+        OrderInterface $order,
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        PaymentMethodInterface $paymentMethod,
+        PaymentInterface $payment,
+        FactoryInterface $stateMachineFactory,
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChoosePaymentMethodHandlerSpec.php:45]: Wrong order, first three arguments should be in order like let function

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChoosePaymentMethodHandlerSpec.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        ;
+    }
+
+    function it_throws_an_exception_if_order_cannot_have_payment_selected(
+        OrderRepositoryInterface $orderRepository,
+        OrderInterface $order,
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        PaymentInterface $payment,
+        FactoryInterface $stateMachineFactory,
+        StateMachineInterface $stateMachine
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChoosePaymentMethodHandlerSpec.php:87]: Like comment above, here and in every places in this file

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChooseShippingMethodHandlerSpec.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    function it_assignes_choosen_shipping_method_to_specified_shipment(
+        OrderRepositoryInterface $orderRepository,
+        OrderInterface $order,
+        ShippingMethodRepositoryInterface $shippingMethodRepository,
+        ShippingMethodInterface $shippingMethod,
+        ShipmentInterface $shipment,
+        ShippingMethodEligibilityCheckerInterface $eligibilityChecker,
+        FactoryInterface $stateMachineFactory,
+        StateMachineInterface $stateMachine
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/ChooseShippingMethodHandlerSpec.php:49]: Similar like comment above

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/CompleteOrderHandlerSpec.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('__invoke', [$completeOrder])
+        ;
+    }
+
+    function it_throws_an_exception_if_order_cannot_be_completed(
+        FactoryInterface $stateMachineFactory,
+        OrderInterface $order,
+        OrderRepositoryInterface $orderRepository,
+        StateMachineInterface $stateMachine
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/CompleteOrderHandlerSpec.php:69]: And here

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+use Sylius\Behat\Client\ResponseCheckerInterface;
+use Sylius\Behat\Service\SharedStorageInterface;
+use Sylius\Component\Core\Model\AddressInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+use Sylius\Component\Core\OrderCheckoutStates;
+use Symfony\Component\BrowserKit\AbstractBrowser;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\VarDumper\VarDumper;
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:26]:
```suggestion
use Symfony\Component\VarDumper\VarDumper;
```

not needed

**Comment by GSadee** [comment]
Closing in favour of https://github.com/Sylius/Sylius/pull/11675

**Pull request closed without merging**